src/items/galleon.rs

```diff
@@ -0,0 +1,318 @@
+        coroutines::{start_coroutine, Coroutine},
+        scene::{Handle, HandleUntyped, Node, RefMut},
+    },
+    prelude::*,
+    rand::gen_range,
+};
+
+use crate::{
+    capabilities,
+    components::{GunlikeAnimation, PhysicsBody, ThrowableItem},
+    nodes::Player,
+    Resources,
+};
+
+pub struct FlyingGalleon {
+    pub sprite: Texture2D,
+    pub pos: Vec2,
+    pub speed: Vec2,
+    pub facing: bool,
+    pub owner_id: u8,
+}
+
+impl FlyingGalleon {
+    pub const SPEED: f32 = 200.;
+    pub const WIDTH: f32 = 425.;
+    pub const HEIGHT: f32 = 390.;
+
+    pub fn new(owner_id: u8) -> FlyingGalleon {
+        let resources = storage::get::<Resources>();
+        let sprite = resources.items_textures["galleon/flying_galleon"];
+
+        let (pos, facing) = Self::start_position();
+        let dir = if facing {
+            Vec2::new(1., 0.)
+        } else {
+            Vec2::new(-1., 0.)
+        };
+
+        FlyingGalleon {
+            sprite,
+            pos,
+            speed: dir * Self::SPEED,
+            facing,
+            owner_id,
+        }
+    }
+
+    pub fn start_position() -> (Vec2, bool) {
+        let resources = storage::get::<Resources>();
+        let map_width =
+            resources.tiled_map.raw_tiled_map.tilewidth * resources.tiled_map.raw_tiled_map.width;
+        let map_height =
+            resources.tiled_map.raw_tiled_map.tileheight * resources.tiled_map.raw_tiled_map.height;
+
+        let facing = gen_range(0, 2) == 0;
+        let pos = Vec2::new(
+            if facing {
+                -Self::WIDTH
+            } else {
+                map_width as f32
+            },
+            gen_range(0., map_height as f32 - Self::HEIGHT),
+        );
+
+        (pos, facing)
+    }
+
+    pub fn update(&mut self) -> bool {
+        self.pos += self.speed * get_frame_time();
+
+        {
+            let resources = storage::get::<Resources>();
+            let map_width = (resources.tiled_map.raw_tiled_map.tilewidth
+                * resources.tiled_map.raw_tiled_map.width) as f32;
+            if self.pos.x + Self::WIDTH < -map_width * 0.5 || self.pos.x > map_width * 1.5 {
+                return false;
+            }
+        }
+
+        for mut player in scene::find_nodes_by_type::<crate::nodes::Player>() {
+            if player.get_hitbox().overlaps(&Rect::new(
+                self.pos.x,
+                self.pos.y,
+                Self::WIDTH,
+                Self::HEIGHT,
+            )) && player.id != self.owner_id
+                && !player.dead
+            {
+                scene::find_node_by_type::<crate::nodes::Camera>()
+                    .unwrap()
+                    .shake_noise(1.0, 10, 1.);
+                {
+                    let mut resources = storage::get_mut::<Resources>();
+                    resources.hit_fxses.spawn(player.body.pos);
+                }
+                let direction = self.pos.x > (player.body.pos.x + 10.);
```

Comment:
Converting hardcoded values to constants makes the code considerably more readable, IMO.

When I approached the code for the first time, all the hardcoded values were... "not helpful" :grin: